### ISSUE TITLE
QMenu spam event.buttons with PySide2

### ISSUE THREAD

**Issue by exor2008** [vispy/vispy#1753]
Hi, when I try to envoke menu by pressing mouse button, I get unexpected behavior.

from PySide2.QtWidgets import QMenu
...

class Canvas(scene.Canvas):
...
def on_mouse_press(self, event):
    print(event.buttons)
    menu = QMenu()
    menu.exec_(event.native.globalPos())

After first press it will print [] - as expected.
Second press - [1]
Third press - [1, 1] and so on.
So, looks like every menu call adds effect of permanent button holding.

**Comment by djhoese** [comment]
Try adding, in both of the event handlers:

```python
if event.handled:
    return

# ... your code ...

event.handled = True
```

My guess is that the press event if never being marked as handled so vispy isn't clearing it out.

**Comment by exor2008** [comment]
Good guess, but nope.
Even after I added your code `event.handled` is allways `False` in `on_mouse_move`.
And it still has `event.button == 2`
Repeated pressing RMB on different view inside same canvas seems to remove the glitch.

**Comment by djhoese** [comment]
> different view inside same canvas

What do you mean by this?

**Comment by exor2008** [comment]
![image](https://user-images.githubusercontent.com/19377172/68961076-872b8100-07e2-11ea-8cf9-cddeb047fdf1.png)

At first everything works fine.
As soon as I envoke menu on view1, event problem appear as I described it above.
If I RMB on view2 everything is getting good again.

**Comment by djhoese** [comment]
@kmuehlbauer @larsoner I'll take any ideas if you guys have some. I'm not sure I'll have time this month to debug this since it will likely involve walking through the event code.

**Comment by kmuehlbauer** [comment]
@djhoese I'll try reproducing. @exor2008 Can you share a minimal runnable example? This will make things easier for me.

**Comment by exor2008** [comment]
@kmuehlbauer sure, here we go
After you RBM on canvas it will allways print 2 `in on_mouse_move`

```
import sys
from PySide2.QtWidgets import QWidget, QApplication, QVBoxLayout, QMenu
from vispy import scene

class DemoScene(QWidget):
    def __init__(self, keys='interactive'):

        super(DemoScene, self).__init__()
        
        #Layout and canvas creation
        box = QVBoxLayout(self)
        self.resize(500,500)
        self.setLayout(box)
        self.canvas = scene.SceneCanvas(keys=keys)
        box.addWidget(self.canvas.native)

        #Connect events
        self.canvas.events.mouse_press.connect(self.on_mouse_press)
        self.canvas.events.mouse_move.connect(self.on_mouse_move)

    def on_mouse_press(self, event):
        if event.button == 2:
            menu = QMenu(self)
            menu.addAction('Action')
            action = menu.exec_(event.native.globalPos())

    def on_mouse_move(self, event):
        print(event.button)


if __name__ == '__main__':
    appQt = QApplication(sys.argv)
    view = DemoScene(keys='interactive')
    view.show()
    appQt.exec_()
```

**Comment by kmuehlbauer** [comment]
@exor2008 Thanks, I'll have a look as soon as possible, hopefully within the next day

**Comment by djhoese** [comment]
I had some time to look at this (procrastinate on real work I should be doing) and I think this is maybe a bug in PyQt. If you take a look at how vispy handles mouse **release** events (https://github.com/vispy/vispy/blob/47f3cd0cf179286774b8ec5d1fbe063a28b8e56b/vispy/app/base.py#L217), you'll see that it detects which button was released and removes it from the cache of mouse information that vispy is holding on to. **However**, when I add some print statements at the top and bottom of this method I **never** see this event for the release of a right click. I see it for left clicks.

FYI here is where the Qt-based backends create the mouse events: https://github.com/vispy/vispy/blob/479c1332d66e276d8b061ed145fdaca72c8b8349/vispy/app/backends/_qt.py#L443

**Comment by djhoese** [comment]
Without the menu I see the release event and this is no longer an issue (it resets the button to `None`).

**Comment by djhoese** [comment]
Newest theory that I'm having trouble testing because I have my right mouse button on my laptop disabled: when you right click and release the release event happens in the context menu and that event never gets to the canvas (to vispy).

**Comment by djhoese** [comment]
Yep, the menu gets the event:

```
class MyMenu(QMenu):
    def mouseReleaseEvent(self, event):
        print("Mouse release: ", event)
        return super(MyMenu, self).mouseReleaseEvent(event)
```

Add that to your code and use that instead of QMenu when you create it and you'll see that the release event goes to the context menu object and *not* to vispy. But of course you can't just pass the mouse event to the canvas (`self.parent().mouseReleaseEvent`) because the context menu needs to actually work with the mouse events when you want to use it.

**Comment by kmuehlbauer** [comment]
@exor2008 The evaluation of @djhoese is correct. The problem is that the `mouse_release` is not handled by the `SceneCanvas` and therefore the EventSystem is stuck in a "mousebutton-pressed" state. 

I was trying to find a way to reset or workaround this to no avail so far. Will do some more testing...

**Comment by exor2008** [comment]
I was trying to emulate releasing RMB right after press with this code:

```
def on_mouse_press(self, event):
    print('press')
    # call menu
    self.events.mouse_release(button=2, buttons=[2],
        last_event=event, press_event=event)


def on_mouse_release(self, release):
    print('release')
```
and I'am actually see 'press' and 'release' on every press and one more 'release' on every release. But it doesn't help with stuck RMB after calling the menu.

As I described it before in the more complex scenario with two views, RMB on view2 helps to unstuck EventSystem.

**Comment by exor2008** [comment]
By the way, the same issue appears, if you try to envoke `QDialog` instead of `QMenu`

**Comment by djhoese** [comment]
@exor2008 If you change your code above to only "fake" the mouse_release if button 2 is pressed and the menu is opened, does it "fix" things?

I'm not sure there is an easy solution to this problem in vispy. We'd have to change how we are handling mouse events for every backend.

You could make custom versions of these QMenu/QDialog classes that do a mouse release if it is the first mouse release after creation. That is the best workaround I can think of right now.

**Comment by kmuehlbauer** [comment]
@djhoese FYI: I've tried that `fake` mouse_release (only for button 2) but this doesn't help, unfortunately.

**Comment by kmuehlbauer** [comment]
@exor2008 Not sure if this will fit your need, but you can do this in `on_mouse_release` instead of `on_mouse_press`

**Comment by exor2008** [comment]
@kmuehlbauer looks like brilliant idea.
I've already remade everything and using `QToolBar` + `QDialog`, but your solution would also fits me very well, thank you, guys.
I think, I shouldn't close an issue, cause actual bug is still a problem, right?

**Comment by kmuehlbauer** [comment]
@exor2008 Yeah, just leave it open. We have still much to think about this...

**Comment by djhoese** [comment]
At the very least this should be documented as a "gotcha"...somewhere.

**Comment by kmuehlbauer** [comment]
@exor2008 Does the problem still persist? If yes, we should decide what to do next. @djhoese is there a place in the docs, where we could at least document it, as you suggested?

**Comment by djhoese** [comment]
I'm not sure the best place to document it. It isn't exactly a FAQ...maybe a new "Gotchas"? Maybe as a sub-section on the FAQ page?

**Comment by mkondratyev85** [comment]
Hello guys! I face the same issue in `vispy` when I work with context menu on a `vispy` widget.

Please look at the following question on stackoverflow. It looks like this it may be related:

https://stackoverflow.com/questions/72386383/how-to-make-widget-receive-mouse-release-event-when-context-menu-appears
